### PR TITLE
Display app names in update page for app updates

### LINKED ISSUE #16272: Don't mention the oC version number when upgrading apps
### Env

Oc 8.0

### Steps to reproduce

1. Go to app manager
2. Enable an app which doesn't make any modification to the database
3. Go to Files

### Expected result

Nothing should happen or if you have to ask the user to upgrade for some reason, then make it clear that this is about an app.

### Actual result

The user is presented with a page telling him that his ownCloud version x.y should be upgraded to version x.y, the exact same version. It is confusing.

### PR DESCRIPTION
**Pull request by PVince81**
Whenever the update page is displayed for apps, show app names instead
of the core update text.

Fixes https://github.com/owncloud/core/issues/16272
- [x] TODO: add unit test
- [x] To discuss: should we move `getIncompatibleApps` and `getAppsNeedingUpgrade` to the `IAppManager` interface ? I'd rather keep them private but "apps.php" is horribly static... @icewind1991 @MorrisJobke @nickvergessen @oparoz what do you think ?

### CODE REVIEW

**Comment by PVince81** [comment]
- [x] TEST: core update: shows core update page
- [x] TEST: app update alone: shows app update page
- [x] TEST: core + app update: shows core update page

**Comment by PVince81** [comment]
- [x] TEST: auto-disable third party apps on core update still works

**Comment by PVince81** [comment]
Setting to "To review" because I'd like to get comments :smile:

**Comment by PVince81** [comment]
Fixes https://github.com/owncloud/core/issues/17437 too, @jancborchardt

**Comment by PVince81** [comment]
@jancborchardt I'll revert and fix separately, it's a different piece of code.

**Comment by PVince81** [comment]
Ok looks like nobody objected the last 1.5 months so I'll keep these methods static.

Time for unit tests then.

**Comment by PVince81** [comment]
Basically it looks like this:
versions-app-update

@jancborchardt

**Comment by oparoz** [comment]
Regarding the static methods. I don't think app devs would have a use for it, unless they write an app manager?

Regarding the design, I think it would be nice if apps could be shown using bullet points. It makes it clearer what is info-text and what is an app.

**Comment by PVince81** [comment]
I checked the code and remembered why I can't move them easily to the app manager: the app manager is missing methods like "isAppCompatible" and "shouldUpgrade". I don't think we want these on the app manager either.

For the bullet points I'll let @jancborchardt decide.
I think until now we never had bullet points on this page for lists. Another list is the one that shows you what apps are getting disabled during a core update.

**Comment by oparoz** [comment]
> Another list is the one that shows you what apps are getting disabled during a core update.

Ah, yes, that needs fixing as well because it only shows the folder name iirc.

**Comment by PVince81** [comment]
@oparoz please raise another ticket for that. Thanks.

**Comment by PVince81** [comment]
Done.

I moved `getIncompatibleApps()` and `getAppsNeedingUpgrade()` to the `AppManager` class.
I also added `AppManager->getAppInfo()` that calls the old API to make it mockable for unit tests.

Please review @icewind1991 @nickvergessen @MorrisJobke @Xenopathic @oparoz

File: core/templates/update.admin.php [hunk 1]
```diff
@@ -1,7 +1,20 @@
 <div class="update" data-productname="<?php p($_['productName']) ?>" data-version="<?php p($_['version']) ?>">
 	<div class="updateOverview">
+		<?php if ($_['isAppsOnlyUpgrade']) { ?>
```
Comment (Xenopathic) [inline core/templates/update.admin.php]: Do app upgrades not occur at the same time as core upgrades? If they do, then we should surely display 'The following apps will be updated' alongside 'ownCloud will be updated'?

Comment (Xenopathic) [inline core/templates/update.admin.php]: As far as I can tell (without doing an actual upgrade), we will see:

> ownCloud will be updated to version x
> 
> External file storage (files_external)
> My awesome app (xenopathic_is_awesome)

IMHO we still need to display 'The following apps will be updated' under 'ownCloud will be updated'

File: core/templates/update.admin.php [hunk 2]
```diff
@@ -17,9 +30,11 @@
 			<?php p($l->t('The theme %s has been disabled.', array($_['oldTheme']))) ?>
 		</div>
 		<?php } ?>
+		<?php if (!$_['isAppsOnlyUpgrade']) { ?>
```
Comment (Xenopathic) [inline core/templates/update.admin.php]: When updating apps, there is just as much chance for a DB schema upgrade to go wrong as there is in core. Same goes for the data folder. IMHO we should display this warning always

File: lib/base.php [hunk 3]
```diff
@@ -375,6 +355,40 @@ public static function checkUpgrade($showTemplate = true) {
 		return false;
 	}
 
+	/**
+	 * Prints the upgrade page
+	 */
+	private static function printUpgradePage() {
+		$systemConfig = \OC::$server->getSystemConfig();
+		$oldTheme = $systemConfig->getValue('theme');
+		$systemConfig->setValue('theme', '');
```
Comment (Xenopathic) [inline lib/base.php]: This is a thing?! :see_no_evil:

Comment (PVince81) [inline lib/base.php]: Possibly... this code was just moved here, I didn't write it.

File: lib/base.php [hunk 4]
```diff
@@ -375,6 +355,40 @@ public static function checkUpgrade($showTemplate = true) {
...
+		$installedVersion = $systemConfig->getValue('version', '0.0.0');
+		$currentVersion = implode('.', OC_Util::getVersion());
+
+		$appManager = \OC::$server->getAppManager();
+
+		$tmpl = new OC_Template('', 'update.admin', 'guest');
+		$tmpl->assign('version', OC_Util::getVersionString());
+
+		// if not a core upgrade, then it's apps upgrade
+		if (version_compare($currentVersion, $installedVersion, '=')) {
```
Comment (PVince81) [inline lib/base.php]: This is the part I added

File: core/templates/update.admin.php (hunk 1 above)
Comment (PVince81) [inline core/templates/update.admin.php]: Yeah, I also thought about this.
I'll try and add them to both pages.

**Comment by PVince81** [comment]
- [ ] TODO: add list of apps to upgrade on both core and app upgrade modes

File: core/templates/update.admin.php (hunk 2 above)
Comment (PVince81) [inline core/templates/update.admin.php]: Hmm, right

**Comment by PVince81** [comment]
- [ ] TODO: keep warning about backups in both modes

File: lib/base.php (hunk 3 above)
Comment (MorrisJobke) [inline lib/base.php]: Yes. Please leave it. It caused a lot of trouble in the past.

**Comment by jancborchardt** [comment]
I agree with @oparoz that it would be nicer to use a list to show the apps. That way they are easily readable and also not center-aligned.

**Comment by PVince81** [comment]
It is already a ul/li list, so I'll switch its CSS style to use bullets instead.
I assume we'll want this for the list of disabled apps too.

- [ ] TODO: add bullets to CSS for "ul" in update page

**Comment by PVince81** [comment]
Done.
I also changed the wording a bit:
versions-app-update1

Note that the main title changes if a core update is required too.

Also fixed:
- [x] BUG: if an app needing update is also incompatible, don't show it in the update list

**Comment by Xenopathic** [comment]
I'm not 100% sure on the 'Apps update required' wording, but I can't think of any alternatives, and the code looks good :+1:

**Comment by MorrisJobke** [comment]
* [ ] disabled 3rdparty apps aren't listed

**Comment by PVince81** [comment]
@MorrisJobke they are listed after the update. Something to be done separately because it might require changing things in the update order.

**Comment by PVince81** [comment]
Note: the original purpose of this PR is to solve the confusion that people have between core update and app update ("why does my ownCloud suddenly wants to update again ?")

**Comment by MorrisJobke** [comment]
Then it is working just fine :+1:

File: core/templates/update.admin.php [hunk 5]
```diff
@@ -1,12 +1,26 @@
 <div class="update" data-productname="<?php p($_['productName']) ?>" data-version="<?php p($_['version']) ?>">
 	<div class="updateOverview">
+		<?php if ($_['isAppsOnlyUpgrade']) { ?>
+		<h2 class="title bold"><?php p($l->t('Apps update required.')); ?></h2>
```
Comment (jancborchardt) [inline core/templates/update.admin.php]: This should say »App update required« (App without s, no dot at the end.)

Comment (MorrisJobke) [inline core/templates/update.admin.php]: https://github.com/owncloud/core/pull/18452

Comment (jancborchardt) [inline core/templates/update.admin.php]: Pull request at https://github.com/owncloud/core/pull/18453

**Pull request merged**